File: client/common/messages.go
package common

import (
	"errors"
	"fmt"
	"os"
	"strings"
)

type BetMessage struct {
	Agency    string
	FirstName string
	LastName  string
	Document  string
	Birthdate string
	Number    string
}

func NewBetMessage(agency string, firstName string, lastName string, document string, birthdate string, number string) *BetMessage {
	return &BetMessage{
		Agency:    agency,
		FirstName: firstName,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthdate,
		Number:    number,
	}
}

func NewBetMessageFromEnv() *BetMessage {
	return &BetMessage{
		Agency:    os.Getenv("CLI_ID"),
		FirstName: os.Getenv("FIRST_NAME"),
		LastName:  os.Getenv("LAST_NAME"),
		Document:  os.Getenv("DOCUMENT"),
		Birthdate: os.Getenv("BIRTH_DATE"),
		Number:    os.Getenv("NUMBER"),
	}
}

func (b *BetMessage) Serialize() string {
	return fmt.Sprintf(
		"BetMessage^%s^%s^%s^%s^%s^%s",
		b.Agency,
		b.FirstName,
		b.LastName,
		b.Document,
		b.Birthdate,
		b.Number,
	)
}

func (b *BetMessage) MatchesBetMessage(s string) bool {
	return matchTag(s, "BetMessage")
}

func (b *BetMessage) Length() int {
	return len(b.Serialize())
}

func DeserializeBetMessage(s string) (*BetMessage, error) {
	parts := strings.Split(s, "^")
	if len(parts) != 7 {
		return nil, errors.New("invalid BetMessage format")
	}

	if parts[0] != "BetMessage" {
		return nil, errors.New("invalid tag: expected BetMessage")
	}

	return &BetMessage{
		Agency:    parts[1],
		FirstName: parts[2],
		LastName:  parts[3],
		Document:  parts[4],
		Birthdate: parts[5],
		Number:    parts[6],
	}, nil
}

type AckMessage struct {
	ProcessedCount string
}

func (a *AckMessage) Serialize() string {
	return fmt.Sprintf("AckMessage^%s", a.ProcessedCount)
}

func MatchesAckMessage(s string) bool {
	return matchTag(s, "AckMessage")
}

func DeserializeAckMessage(s string) (*AckMessage, error) {
	parts := strings.Split(s, "^")
	if len(parts) != 2 {
		return nil, errors.New("invalid AckMessage format")
	}
	if parts[0] != "AckMessage" {
		return nil, errors.New("invalid tag: expected AckMessage")
	}
	return &AckMessage{
		ProcessedCount: parts[1],
	}, nil
}

func matchTag(s, expected string) bool {
	parts := strings.SplitN(s, "^", 2)
	return len(parts) > 0 && parts[0] == expected
}

File: server/common/messages.py
from abc import ABC, abstractmethod

########################################################
# ABSTRACT MESSAGE
########################################################


class Message(ABC):
    _FIELD_DELIM = "^"
    _RECORD_DELIM = "~"

    @classmethod
    def get_tag(cls) -> str:
        return cls.__name__

    @classmethod
    @abstractmethod
    def matches(cls, message_encoded: bytes) -> bool: ...

    @classmethod
    @abstractmethod
    def deserialize(cls, message_encoded: bytes) -> "Message": ...

    @abstractmethod
    def serialize(self) -> bytes: ...


########################################################
# MESSAGES RECEIVED
########################################################


class BetMessage(Message):
    def __init__(
        self,
        agency: str,
        first_name: str,
        last_name: str,
        document: str,
        birthdate: str,
        number: str,
    ):
        self.agency: str = agency
        self.first_name: str = first_name
        self.last_name: str = last_name
        self.document: str = document
        self.birthdate: str = birthdate
        self.number: str = number

    @classmethod
    def matches(cls, message_encoded: bytes) -> bool:
        string = message_encoded.decode("utf-8")
        parts = string.split(cls._FIELD_DELIM)
        return parts and parts[0] == cls.get_tag()

    @classmethod
    def deserialize(cls, message_encoded: bytes) -> "BetMessage":
        string = message_encoded.decode("utf-8")
        fields: list = string.split(cls._FIELD_DELIM)

        if len(fields) != 7 or fields[0] != cls.get_tag():
            raise ValueError(f"Invalid {cls.get_tag()} format: {string!r}")

        return cls(*fields[1:])

    def serialize(self) -> bytes:
        raise NotImplementedError


class BetBatchMessage(Message):
    def __init__(self, bet_messages: list[BetMessage]):
        self.bet_messages: list[BetMessage] = bet_messages

    @classmethod
    def matches(cls, message_encoded: bytes) -> bool:
        string = message_encoded.decode("utf-8")
        parts = string.split(cls._RECORD_DELIM)
        return parts and parts[0] == cls.get_tag()

    @classmethod
    def deserialize(cls, message_encoded: bytes) -> "BetBatchMessage":
        string = message_encoded.decode("utf-8")
        records: list[str] = string.split(cls._RECORD_DELIM)

        if len(records) < 2 or records[0] != cls.get_tag():
            raise ValueError(f"Invalid {cls.get_tag()} format: {string!r}")

        bet_messages: list[BetMessage] = [
            BetMessage.deserialize(record_str.encode("utf-8"))
            for record_str in records[1:]
        ]
        return cls(bet_messages=bet_messages)

    def serialize(self) -> bytes:
        raise NotImplementedError


########################################################
# MESSAGES SENT
########################################################


class AckMessage(Message):
    def __init__(self, success: bool):
        super().__init__()
        self.success: bool = success

    @classmethod
    def matches(cls, message_encoded: bytes) -> bool:
        raise NotImplementedError

    @classmethod
    def deserialize(cls, message_encoded: bytes) -> "AckMessage":
        raise NotImplementedError

    def serialize(self) -> bytes:
        string = self._FIELD_DELIM.join(
            [
                self.get_tag(),
                str(self.success),
            ]
        )
        return string.encode("utf-8")
